provider/ec2/image.go

```diff
@@ -44,20 +37,53 @@ func filterImages(images []*imagemetadata.ImageMetadata, ic *instances.InstanceC
 
 // findInstanceSpec returns an InstanceSpec satisfying the supplied instanceConstraint.
 func findInstanceSpec(
+	controller bool,
 	allImageMetadata []*imagemetadata.ImageMetadata,
+	instanceTypes []instances.InstanceType,
 	ic *instances.InstanceConstraint,
 ) (*instances.InstanceSpec, error) {
 	logger.Debugf("received %d image(s)", len(allImageMetadata))
-	// If the instance type is set, don't also set a default CPU power
-	// as this is implied.
-	cons := ic.Constraints
-	if cons.CpuPower == nil && (cons.InstanceType == nil || *cons.InstanceType == "") {
-		ic.Constraints.CpuPower = instances.CpuPower(defaultCpuPower)
+	if controller {
+		ic.Constraints = withDefaultControllerConstraints(ic.Constraints)
+	} else {
+		ic.Constraints = withDefaultNonControllerConstraints(ic.Constraints)
 	}
 	suitableImages := filterImages(allImageMetadata, ic)
 	logger.Debugf("found %d suitable image(s)", len(suitableImages))
 	images := instances.ImageMetadataToImages(suitableImages)
-
-	instanceTypes := ec2instancetypes.RegionInstanceTypes(ic.Region)
 	return instances.FindInstanceSpec(images, ic, instanceTypes)
 }
+
+// withDefaultControllerConstraints returns the given constraints,
+// updated to choose a default instance type appropriate for a
+// controller machine. We use this only if the user does not specify
+// any constraints that would otherwise control the instance type
+// selection.
+//
+// At the time of writing, this will choose
+//   - t2.medium, for VPC
```

Comment:
This is probably all fine. Just to check - Uros has a concern about controllers being CPU bound. Is t2.medium sufficient? Is it correct not to require an instance with a specified minimum cpu-power figure?